guides/2021-05-02-build-flags-backwards-compatibility/en.markdown

```diff
@@ -0,0 +1,148 @@
+sometimes break downstream dependencies given that in the case of `go`, when you're importing a third party module,
+you're forced to build your project with the latest version of the complete dependency tree. 
+
+This guide explains how to deal with situations where you want to use new features of recent versions of `go`
+and at the same time you don't want to force downstream dependecies to upgrade. In this case we'll be using 
+conditional compilation through `go build` tags in a real case scenario by using `go` 1.15 and 1.16 new `io.FS`
+package respectively. 
+
+
+### A simple go 1.15 program using ioutil.Discard
+
+Let's start by checking our current `go` 1.15 version:
+
+{{{ step "go115version" }}}
+
+Now, we'll also check that we have `go` 1.16 installed as well:
+
+{{{ step "go116version" }}}
+
+
+We'll start by setting `go` 1.15 as the default working version:
+
+{{{ step "go115default" }}}
+
+Start by a new `public` module:
+
+{{{ step "public_init" }}}
+
+Now, we'll upload a simple `go` program that uses 1.15 `ioutil.Discard` in a public function:
+
+Create an initial version of the `{{{ .public_func }}}` in `{{{ .public_go }}}`:
+
+{{{ step "public_go_initial" }}}
+
+Commit and push this initial version:
+
+{{{ step "public_initial_commit" }}}
+
+### The `{{{ .gopher }}}` module
+
+Now create a `{{{ .gopher }}}` module to try out the `{{{ .public }}}` module. Unlike
+the `{{{ .public }}}` module, you will not publish the `{{{ .gopher }}}` module; it
+will be local only:
+
+{{{ step "gopher_init" }}}
+
+Create an initial version of a `main` package that uses the previous public module, in `{{{ .gopher_go }}}`:
+
+{{{ step "gopher_go_initial" }}}
+
+
+Now, we'll run the `{{{ .gopher }}}` module `main` package:
+
+{{{ step "gopher_run" }}}
+
+
+### Bumping the `{{{ .public }}}` dependency to `go` 1.16
+
+`Go` 1.16 has been released, and as good developers we want to refactor our
+code so it uses the new `io.Discard` variable instead of the deprecated `ioutil` one.
+
+
+First, we set our `go` version to 1.16:
+
+{{{ step "go116default" }}}
+
+Now, we change our original `{{{ .public }}}` module to use the new variable.
+
+{{{ step "public_bump_discard"}}}
+
+
+Now we're good to release a new version of our `{{{ .public }}}` module
+using the `go` 1.16 new package:
+
+{{{ step "public_bump_commit"}}}
+
+
+Let's go back to our `{{{ .gopher }}}` module in `go` 1.15 and try to fetch
+the latest version of the `{{{ .public }}}` dependecy and see what happens
+
+{{{ step "go115default1" }}}
+
+{{{ step "gopher_update_fail" }}}
+
+
+As you can see, when trying to bump our `{{[ .gopher ]}}` project, we got an 
+error because in our case, we're still using `go` 1.15 in our project which
+doesn't support the new `io.Discard` package.
+
+How do we handle these situations where we shouldn't force our clients to update?
+The right approach to tackle this is by using `build tag` so our `{{{ .public }}}`
```

Comment:
The special comments in the files are officially called build constraints (I'm sure I'm guilty of being sloppy and using tag when I mean constraint):

https://pkg.go.dev/go/build#hdr-Build_Constraints

Indeed, linking to that package might be a good idea for the full docs, or else you could have a quick tangent to describe the different types of build constraints that exist?